# Practice notebook for confidence intervals using NHANES data

This notebook will give you the opportunity to practice working with confidence intervals using the NHANES data.

You can enter your code into the cells that say "enter your code here", and you can type responses to the questions into the cells that say "Type Markdown and Latex".

Note that most of the code that you will need to write below is very similar to code that appears in the case study notebook.  You will need to edit code from that notebook in small ways to adapt it to the prompts below.

To get started, we will use the same module imports and read the data in the same way as we did in the case study:

In [2]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns
import statsmodels.api as sm

df = pd.read_csv("nhanes_2015_2016.csv")

## Question 1

Restrict the sample to women between 35 and 50, then use the marital status variable [DMDMARTL](https://wwwn.cdc.gov/Nchs/Nhanes/2015-2016/DEMO_I.htm#DMDMARTL) to partition this sample into two groups - women who are currently married, and women who are not currently married.  Within each of these groups, calculate the proportion of women who have completed college.  Calculate 95% confidence intervals for each of these proportions.

In [3]:
# enter your code here
df.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
3       1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4       1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  
3     37.7     38.3     110.1     2.0  
4     36.0     27.2      80.4     2.0  

[5 rows x 28 columns]

In [15]:
mask = (da['RIDAGEYR'] >= 35) & (da['RIDAGEYR'] <= 50) & (da['RIAGENDR'] == 2)

In [16]:
female_35_50 = df[mask]

In [17]:
female_35_50.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
4   83736     2.0     1.0     1.0       2         2        42         4   
34  83799     NaN     NaN     NaN       2         2        37         2   
50  83828     1.0     NaN     2.0       2         2        39         1   
52  83832     2.0     1.0     4.0       2         2        50         1   
55  83837     2.0     2.0     NaN       2         2        45         1   

    DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
4        1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   
34       1.0       4.0  ...   110.0    72.0   66.6  161.6    25.5     NaN   
50       2.0       3.0  ...   100.0    62.0   71.3  162.0    27.2    36.8   
52       2.0       1.0  ...     NaN     NaN  105.9  157.7    42.6    29.2   
55       1.0       2.0  ...   114.0    68.0   77.5  148.3    35.2    30.5   

    BMXARML  BMXARMC  BMXWAIST  HIQ210  
4      36.0     27.2      80.4     2.0  
34      NaN      NaN       NaN     2.0  
50     34.6     29.1      94.6     NaN  
52     35.0     40.7     129.1     NaN  
55     34.0     34.4     107.6     2.0  

[5 rows x 28 columns]

### DMDEDUC2
##### less than 9th grade education, 9-11th grade education (includes 12th grade and no diploma), High school graduate/GED, some college or associates (AA) degree, and college graduate or higher

In [18]:
female_35_50['DMDEDUC2'] = female_35_50['DMDEDUC2'].replace([1.0,2.0,3.0,4.0,5.0,7.0, 9.0], ['incomplete_college', 'incomplete_college', 'incomplete_college', 'incomplete_college', 'complete_college', 'incomplete_college', 'incomplete_college'])

In [19]:
female_35_50['DMDMARTL'] = female_35_50['DMDMARTL'].replace([1.0,2.0,3.0,4.0,5.0,6.0,77.0, 99.0], ['married', 'not_married', 'not_married', 'not_married', 'not_married', 'not_married', 'not_married', 'not_married'])

In [20]:
female_35_50

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
4     83736     2.0     1.0     1.0       2         2        42         4   
34    83799     NaN     NaN     NaN       2         2        37         2   
50    83828     1.0     NaN     2.0       2         2        39         1   
52    83832     2.0     1.0     4.0       2         2        50         1   
55    83837     2.0     2.0     NaN       2         2        45         1   
...     ...     ...     ...     ...     ...       ...       ...       ...   
5658  93568     1.0     NaN     1.0       1         2        46         3   
5685  93612     2.0     2.0     NaN       2         2        36         5   
5689  93619     NaN     NaN     NaN       2         2        44         5   
5721  93676     1.0     NaN     2.0       2         2        35         4   
5724  93682     NaN     NaN     NaN       2         2        41         5   

      DMDCITZN            DMDEDUC2  ... BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  \
4          1.0  incomplete_college  ...  114.0    54.0   55.2  164.9    20.3   
34         1.0  incomplete_college  ...  110.0    72.0   66.6  161.6    25.5   
50         2.0  incomplete_college  ...  100.0    62.0   71.3  162.0    27.2   
52         2.0  incomplete_college  ...    NaN     NaN  105.9  157.7    42.6   
55         1.0  incomplete_college  ...  114.0    68.0   77.5  148.3    35.2   
...        ...                 ...  ...    ...     ...    ...    ...     ...   
5658       1.0  incomplete_college  ...  128.0    84.0   85.5  152.4    36.8   
5685       1.0    complete_college  ...  112.0    74.0   65.7  162.4    24.9   
5689       2.0  incomplete_college  ...  118.0    78.0   51.9  149.3    23.3   
5721       1.0    complete_college  ...  114.0    76.0   92.2  161.7    35.3   
5724       1.0    complete_college  ...  122.0    84.0   58.2  166.9    20.9   

      BMXLEG  BMXARML  BMXARMC  BMXWAIST  HIQ210  
4       37.4     36.0     27.2      80.4     2.0  
34       NaN      NaN      NaN       NaN     2.0  
50      36.8     34.6     29.1      94.6     NaN  
52      29.2     35.0     40.7     129.1     NaN  
55      30.5     34.0     34.4     107.6     2.0  
...      ...      ...      ...       ...     ...  
5658    26.0     33.5     34.7     116.5     2.0  
5685    37.2     36.0     29.8      89.0     2.0  
5689    31.9     32.0     26.2      81.4     1.0  
5721    41.5     37.5     38.9     110.9     2.0  
5724    37.1     35.3     26.9      80.8     2.0  

[787 rows x 28 columns]

In [101]:
female_35_50.groupby(['DMDMARTL', 'DMDEDUC2']).agg({'DMDEDUC2': [np.size]})

DMDEDUC2
                                   size
DMDMARTL    DMDEDUC2                   
married     complete_college        162
            incomplete_college      287
not_married complete_college         72
            incomplete_college      266

In [108]:
#female_35_50.groupby(['DMDMARTL', 'DMDEDUC2']).agg({'DMDEDUC2': [np.size], 'DMDMARTL': [np.size, len]})# / len(female_35_50)

In [37]:
female_35_50_MC = pd.crosstab(female_35_50['DMDMARTL'], female_35_50['DMDEDUC2'])

In [39]:
female_35_50_MC

DMDEDUC2     complete_college  incomplete_college
DMDMARTL                                         
married                   162                 287
not_married                72                 266

In [47]:
total = female_35_50_MC.sum(axis=1).sum()
married_women = female_35_50_MC.sum(axis=1)[0]
unmarried_women = female_35_50_MC.sum(axis=1)[1]
unmarried_women_col = female_35_50_MC.iloc[1][0]
married_women_col = female_35_50_MC.iloc[0][0]


In [48]:
total

787

In [49]:
married_women, unmarried_women

(449, 338)

In [50]:
married_women_col, unmarried_women_col

(162, 72)

In [51]:
p_mw_in_col = married_women_col / married_women
p_umw_in_col = unmarried_women_col / unmarried_women

p_mw_in_col, p_umw_in_col

(0.36080178173719374, 0.21301775147928995)

In [52]:
np.sqrt(p_mw_in_col * (1 - p_mw_in_col) / married_women_col)

0.03773067783095193

In [53]:
np.sqrt(p_umw_in_col * (1 - p_umw_in_col) / unmarried_women_col)

0.04825297760443814


### Confidence Interval Formula
#### lcb = p - 1.96 * np.sqrt(p * (1 - p) / n)  
#### ucb = p + 1.96 * np.sqrt(p * (1 - p) / n) 

In [54]:
lcb_mw = p_mw_in_col - 1.96 * np.sqrt(p_mw_in_col * (1 - p_mw_in_col) / married_women)
ucb_mw = p_mw_in_col + 1.96 * np.sqrt(p_mw_in_col * (1 - p_mw_in_col) / married_women) 

In [55]:
lcb_mw, ucb_mw

(0.3163811208674688, 0.4052224426069187)

In [56]:
sm.stats.proportion_confint(married_women_col, married_women) 

(0.31638193710753626, 0.4052216263668512)

In [57]:
lcb_umw = p_umw_in_col - 1.96 * np.sqrt(p_umw_in_col * (1 - p_umw_in_col) / unmarried_women)
ucb_umw = p_umw_in_col + 1.96 * np.sqrt(p_umw_in_col * (1 - p_umw_in_col) / unmarried_women) 

In [58]:
lcb_umw, ucb_umw

(0.1693673655848136, 0.25666813737376626)

In [59]:
sm.stats.proportion_confint(unmarried_women_col, unmarried_women) 

(0.16936816767089768, 0.2566673352876822)

__Q1a.__ Identify which of the two confidence intervals is wider, and explain why this is the case. 

In [60]:
 ci_range1 = ucb_umw -lcb_umw
 ci_range1

0.08730077178895265

In [61]:
ci_range2 = ucb_mw - lcb_mw
ci_range2

0.08884132173944992

### Though, not a great difference from the unmarried women, the confidence interval of the married women is wider. This most likely is due to lesser sample size because there is inverse square root relationship between confidence intervals and sample sizes. Also a higher standard error would translate to a higher confidence interval.

__Q1b.__ Write 1-2 sentences summarizing these findings for an audience that does not know what a confidence interval is (the goal here is to report the substance of what you learned about how marital status and educational attainment are related, not to teach a person what a confidence interval is).

There is 95% certainty that out of every 100 women, if unmarried, 78 to 82 would complete college and if married, only 66 to 71 would complete college. 
From this, we can deduce that unmarried women are more likely to go further in their education.

## Question 2

Construct 95% confidence intervals for the proportion of smokers who are female, and for the proportion of smokers who are male.  Then construct a 95% confidence interval for the difference between these proportions.

In [62]:
# enter your code here
df.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
3       1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4       1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  
3     37.7     38.3     110.1     2.0  
4     36.0     27.2      80.4     2.0  

[5 rows x 28 columns]

In [63]:
df["SMQ020x"] = df.SMQ020.replace({1: "Yes", 2: "No", 7: np.nan, 9: np.nan})  
df["RIAGENDRx"] = df.RIAGENDR.replace({1: "Male", 2: "Female"})

In [64]:
smokers = df[["SMQ020x", "RIAGENDRx"]].dropna()
pd.crosstab( smokers.RIAGENDRx, smokers.SMQ020x)

SMQ020x      No   Yes
RIAGENDRx            
Female     2066   906
Male       1340  1413

In [65]:
smokers_prop = smokers.groupby(smokers.RIAGENDRx).agg({"SMQ020x": [lambda x: np.mean(x=="Yes"), np.size]})
smokers_prop.columns = ["Proportion", "Total_n"] 
smokers_prop

Proportion  Total_n
RIAGENDRx                     
Female       0.304845     2972
Male         0.513258     2753

In [66]:
p_female = smokers_prop.Proportion.Female 
n_female = smokers_prop.Total_n.Female 
se_female = np.sqrt(p_female * (1 - p_female) / n_female)
print(se_female)

p_male = smokers_prop.Proportion.Male 
n_male = smokers_prop["Total_n"].Male 
se_male = np.sqrt(p_male * (1 - p_male) / n_male)
print(se_male)

0.008444152146214435
0.009526078653689868


In [67]:

lcb_female = p_female - 1.96 * np.sqrt(p_female * (1 - p_female) / n_female)  
ucb_female = p_female + 1.96 * np.sqrt(p_female * (1 - p_female) / n_female)  
print(lcb_female, ucb_female)

0.288294683866098 0.32139576027925865


In [68]:

lcb_male = p_male - 1.96 * np.sqrt(p_male * (1 - p_male) / n_male)  
ucb_male = p_male + 1.96 * np.sqrt(p_male * (1 - p_male) / n_male)  
print(lcb_male, ucb_male)

0.49458714955108174 0.531929377873546


In [69]:
se_diff = np.sqrt(se_female**2 + se_male**2)
se_diff

0.012729881381407434

In [70]:
d = smokers_prop.Proportion.Female - smokers_prop.Proportion.Male
lcb = d - 2*se_diff
ucb = d + 2*se_diff
print(lcb, ucb)

-0.2338728044024504 -0.18295327887682067


__Q2a.__ Discuss why it may be relevant to report the proportions of smokers who are female and male, and contrast this to reporting the proportions of males and females who smoke.


__Q2b.__ How does the width of the confidence interval for the difference of the two proportions compare to the widths of the confidence intervals for each proportion separately?

### Ans: 

The 95% confidence interval for the difference of the two proportions  tells us that any value for the difference of population proportions (between females and males) lying between -0.233 and -0.183 is consistent with the observed data

## Question 3

Construct a 95% interval for height ([BMXHT](https://wwwn.cdc.gov/Nchs/Nhanes/2015-2016/BMX_I.htm#BMXHT)) in centimeters.  Then convert height from centimeters to inches by dividing by 2.54, and construct a 95% confidence interval for height in inches.  Finally, convert the endpoints (the lower and upper confidence limits) of the confidence interval from inches to back to centimeters   

In [71]:
# enter your code here
df.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BMXWT  BMXHT  BMXBMI  BMXLEG  BMXARML  BMXARMC  \
0       1.0       5.0  ...   94.8  184.5    27.8    43.3     43.6     35.9   
1       2.0       3.0  ...   90.4  171.4    30.8    38.0     40.0     33.2   
2       1.0       3.0  ...   83.4  170.1    28.8    35.6     37.0     31.0   
3       1.0       5.0  ...  109.8  160.9    42.4    38.5     37.7     38.3   
4       1.0       4.0  ...   55.2  164.9    20.3    37.4     36.0     27.2   

   BMXWAIST  HIQ210  SMQ020x  RIAGENDRx  
0     101.1     2.0      Yes       Male  
1     107.9     NaN      Yes       Male  
2     116.5     2.0      Yes       Male  
3     110.1     2.0       No     Female  
4      80.4     2.0       No     Female  

[5 rows x 30 columns]

In [72]:
heights = df['BMXHT']
height_cm_mean = np.mean(heights)
height_cm_std = np.std(heights)

In [73]:
height_inches_mean = np.mean(heights / 2.54)
height_inches_std = np.std(heights / 2.54)

In [74]:
height_cm_mean, height_cm_std

(166.1428344791116, 10.078375319383921)

In [75]:
len(heights)

5735

In [78]:
lcb_height_cm = height_cm_mean - 1.96 * height_cm_std / np.sqrt(len(heights))
ucb_height_cm = height_cm_mean + 1.96 * height_cm_std / np.sqrt(len(heights))

In [79]:
lcb_height_cm, ucb_height_cm

(165.88199105792793, 166.40367790029526)

In [81]:
sm.stats.DescrStatsW(da['BMXHT'].dropna()).zconfint_mean()

(165.88055125872887, 166.40511769949427)

In [82]:
height_inches = da['BMXHT'].dropna() / 2.54

In [83]:
sm.stats.DescrStatsW(height_inches).zconfint_mean()

(65.30730364516884, 65.51382586594264)

In [207]:
#sm.stats.DescrStatsW.zconfint_mean(da['BMXHT'].values, alpha=0.05)

In [84]:
lcb_height_inches = height_inches_mean - 1.96 * height_inches_std / np.sqrt(len(heights))
ucb_height_inches = height_inches_mean + 1.96 * height_inches_std / np.sqrt(len(heights))

In [85]:
height_inches_mean, height_inches_std

(65.41056475555574, 3.9678642989700483)

In [86]:
lcb_height_inches, ucb_height_inches

(65.30787049524722, 65.51325901586426)

__Q3a.__ Describe how the confidence interval constructed in centimeters relates to the confidence interval constructed in inches.

### As illustrated below, the confidence intervals are the same if converted between units i.e centimeters and inches

In [150]:
lcb_height_cm, ucb_height_cm 

(165.88199105792793, 166.40367790029526)

In [151]:
lcb_height_inches, ucb_height_inches

(65.30787049524722, 65.51325901586426)

In [154]:
lcb_height_inches * 2.54, ucb_height_inches *2.54

(165.88199105792793, 166.40367790029524)

## Question 4

Partition the sample based on 10-year age bands, i.e. the resulting groups will consist of people with ages from 18-28, 29-38, etc. Construct 95% confidence intervals for the difference between the mean BMI for females and for males within each age band.

In [90]:
df['RIDAGEYR'].unique().min(), df['RIDAGEYR'].unique().max()

(18, 80)

In [183]:
# enter your code here
#da['Age_grouped'] = pd.cut(da['RIDAGEYR'], bins=list(range(18, 88, 10)))

In [91]:
# Calculate the mean, SD, and sample size for BMI within age/gender groups
df["agegrp"] = pd.cut(df.RIDAGEYR, [18, 30, 40, 50, 60, 70, 80])
pr = df.groupby(["agegrp", "RIAGENDRx"]).agg({"BMXBMI": [np.mean, np.std, np.size]}).unstack()

# Calculate the Standard Error of Mean for females and for males within each age band
pr["BMXBMI", "sem", "Female"] = pr["BMXBMI", "std", "Female"] / np.sqrt(pr["BMXBMI", "size", "Female"]) 
pr["BMXBMI", "sem", "Male"] = pr["BMXBMI", "std", "Male"] / np.sqrt(pr["BMXBMI", "size", "Male"]) 

# Calculate the mean difference of BMI between females and males within each age band, also  calculate
# its SE and the lower and upper limits of its 95% CI.
pr["BMXBMI", "mean_diff", ""] = pr["BMXBMI", "mean", "Female"] - pr["BMXBMI", "mean", "Male"]
pr["BMXBMI", "sem_diff", ""] = np.sqrt(pr["BMXBMI", "sem", "Female"]**2 + pr["BMXBMI", "sem", "Male"]**2) 
pr["BMXBMI", "lcb_diff", ""] = pr["BMXBMI", "mean_diff", ""] - 1.96 * pr["BMXBMI", "sem_diff", ""] 
pr["BMXBMI", "ucb_diff", ""] = pr["BMXBMI", "mean_diff", ""] + 1.96 * pr["BMXBMI", "sem_diff", ""]

In [92]:
pr

BMXBMI                                                         \
                mean                  std             size              sem   
RIAGENDRx     Female       Male    Female      Male Female   Male    Female   
agegrp                                                                        
(18, 30]   28.123881  27.391822  7.745893  6.649440  609.0  544.0  0.313879   
(30, 40]   30.325586  29.611726  8.315608  6.622412  474.0  458.0  0.381949   
(40, 50]   31.160643  29.724623  8.076195  6.407076  502.0  401.0  0.360458   
(50, 60]   30.743777  29.231486  7.575848  5.914373  470.0  454.0  0.349448   
(60, 70]   31.074828  29.392488  7.604514  5.933307  441.0  437.0  0.362120   
(70, 80]   29.138213  27.957692  6.284968  4.974855  410.0  402.0  0.310392   

                                                             
                    mean_diff  sem_diff  lcb_diff  ucb_diff  
RIAGENDRx      Male                                          
agegrp                                                       
(18, 30]   0.285092  0.732059  0.424026 -0.099032  1.563150  
(30, 40]   0.309445  0.713861  0.491570 -0.249616  1.677338  
(40, 50]   0.319954  1.436019  0.481976  0.491347  2.380692  
(50, 60]   0.277575  1.512291  0.446275  0.637591  2.386991  
(60, 70]   0.283829  1.682340  0.460097  0.780550  2.584130  
(70, 80]   0.248123  1.180521  0.397377  0.401662  1.959380

__Q4a.__ How do the widths of these confidence intervals differ?  Provide an explanation for any substantial diferences in the confidence interval widths that you see.

## Question 5

Construct a 95% confidence interval for the first and second systolic blood pressure measures, and for the difference between the first and second systolic blood pressure measurements within a subject.

In [185]:
# enter code here
bp = pd.read_csv('blood_pressure.csv')
bp.head()

patient   sex agegrp  bp_before  bp_after
0        1  Male  30-45        143       153
1        2  Male  30-45        163       170
2        3  Male  30-45        153       168
3        4  Male  30-45        153       142
4        5  Male  30-45        146       141

In [208]:
first_bp = bp['bp_before']
second_bp = bp['bp_after']

first_bp_mean = np.mean(first_bp)
first_bp_std = np.std(first_bp)

second_bp_mean = np.mean(second_bp)
second_bp_std = np.std(second_bp)

n = len(bp)


In [209]:
first_sbp = df['BPXSY1']
second_sbp = df['BPXSY2']

first_sbp_mean = np.mean(first_sbp)
first_sbp_std = np.std(first_sbp)

second_sbp_mean = np.mean(second_sbp)
second_sbp_std = np.std(second_sbp)

sbp_length = len(df)


120

In [194]:

first_bp_mean, second_bp_mean

(156.45, 151.35833333333332)

In [112]:
first_sbp_mean, second_sbp_mean

(125.08461396037771, 124.78301716350497)

In [113]:
first_sbp_std, second_sbp_std

(18.479161697371598, 18.525338021233786)

In [114]:
lcb_first_sbp = first_sbp_mean - 1.96 * first_sbp_std / np.sqrt(sbp_length)
ucb_first_sbp = first_sbp_mean + 1.96 * first_sbp_std / np.sqrt(sbp_length)

In [115]:
lcb_second_sbp = second_sbp_mean - 1.96 * second_sbp_std / np.sqrt(sbp_length)
ucb_second_sbp = second_sbp_mean + 1.96 * second_sbp_std / np.sqrt(sbp_length)

In [109]:
df.columns

Index(['SEQN', 'ALQ101', 'ALQ110', 'ALQ130', 'SMQ020', 'RIAGENDR', 'RIDAGEYR',
       'RIDRETH1', 'DMDCITZN', 'DMDEDUC2', 'DMDMARTL', 'DMDHHSIZ', 'WTINT2YR',
       'SDMVPSU', 'SDMVSTRA', 'INDFMPIR', 'BPXSY1', 'BPXDI1', 'BPXSY2',
       'BPXDI2', 'BMXWT', 'BMXHT', 'BMXBMI', 'BMXLEG', 'BMXARML', 'BMXARMC',
       'BMXWAIST', 'HIQ210', 'SMQ020x', 'RIAGENDRx', 'agegrp'],
      dtype='object')

In [110]:
df.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BMXHT  BMXBMI  BMXLEG  BMXARML  BMXARMC  BMXWAIST  \
0       1.0       5.0  ...  184.5    27.8    43.3     43.6     35.9     101.1   
1       2.0       3.0  ...  171.4    30.8    38.0     40.0     33.2     107.9   
2       1.0       3.0  ...  170.1    28.8    35.6     37.0     31.0     116.5   
3       1.0       5.0  ...  160.9    42.4    38.5     37.7     38.3     110.1   
4       1.0       4.0  ...  164.9    20.3    37.4     36.0     27.2      80.4   

   HIQ210  SMQ020x  RIAGENDRx    agegrp  
0     2.0      Yes       Male  (60, 70]  
1     NaN      Yes       Male  (50, 60]  
2     2.0      Yes       Male  (70, 80]  
3     2.0       No     Female  (50, 60]  
4     2.0       No     Female  (40, 50]  

[5 rows x 31 columns]

In [117]:
lcb_first_sbp, ucb_first_sbp

(124.60634562793352, 125.5628822928219)

In [119]:

sm.stats.DescrStatsW(df['BPXSY1'].dropna()).zconfint_mean()

(124.59174272058787, 125.57748520016754)

In [122]:
lcb_second_sbp, ucb_second_sbp

(124.30355371876219, 125.26248060824774)

In [123]:

sm.stats.DescrStatsW(df['BPXSY2'].dropna()).zconfint_mean()

(124.29493306967777, 125.27110125733216)

In [127]:
sbp_mean_diff = first_sbp_mean - second_sbp_mean
sbp_mean_diff

0.30159679687274377

In [125]:
sem_first_sbp = first_sbp_std / np.sqrt(sbp_length)
sem_second_sbp = second_sbp_std / np.sqrt(sbp_length)

In [129]:
sbp_sem_diff_sbp = np.sqrt(sem_first_sbp**2 + sem_second_sbp**2)
sbp_sem_diff_sbp

0.3455199803423725

In [130]:
lcb_mean_diff_sbp = sbp_mean_diff - 2*sbp_sem_diff_sbp
ucb_mean_diff_sbp = sbp_mean_diff + 2*sbp_sem_diff_sbp
lcb_mean_diff_sbp, ucb_mean_diff_sbp

(-0.3894431638120013, 0.9926367575574888)

__Q5a.__ Based on these confidence intervals, would you say that a difference of zero between the population mean values of the first and second systolic blood pressure measures is consistent with the data?


__Q5b.__ Discuss how the width of the confidence interval for the within-subject difference compares to the widths of the confidence intervals for the first and second measures.

## Question 6

Construct a 95% confidence interval for the mean difference between the average age of a smoker, and the average age of a non-smoker.

In [234]:
# insert your code here
da.head()


SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BMXBMI  BMXLEG  BMXARML  BMXARMC  BMXWAIST  \
0       1.0       5.0  ...    27.8    43.3     43.6     35.9     101.1   
1       2.0       3.0  ...    30.8    38.0     40.0     33.2     107.9   
2       1.0       3.0  ...    28.8    35.6     37.0     31.0     116.5   
3       1.0       5.0  ...    42.4    38.5     37.7     38.3     110.1   
4       1.0       4.0  ...    20.3    37.4     36.0     27.2      80.4   

   HIQ210  SMQ020x  RIAGENDRx  Age_grouped    agegrp  
0     2.0      Yes       Male     (58, 68]  (60, 70]  
1     NaN      Yes       Male     (48, 58]  (50, 60]  
2     2.0      Yes       Male     (68, 78]  (70, 80]  
3     2.0       No     Female     (48, 58]  (50, 60]  
4     2.0       No     Female     (38, 48]  (40, 50]  

[5 rows x 32 columns]

In [297]:
mean_age_smoker = da.groupby('SMQ020x').apply(lambda x: x['RIDAGEYR'].mean())[1]
mean_age_nonsmoker = da.groupby('SMQ020x').apply(lambda x: x['RIDAGEYR'].mean())[0]

In [310]:
smokers = da[da['SMQ020x'] == 'Yes']['RIDAGEYR']
non_smokers = da[da['SMQ020x'] == 'No']['RIDAGEYR']

In [312]:
sm.stats.DescrStatsW(smokers).zconfint_mean()

(51.38591951147112, 52.80726720694198)

In [311]:
sm.stats.DescrStatsW(non_smokers).zconfint_mean()

(44.637087398171616, 45.882583770354515)

__Q6a.__ Use graphical and numerical techniques to compare the variation in the ages of smokers to the variation in the ages of non-smokers.  

In [321]:
# insert your code here

da.groupby(["SMQ020x", "Age_grouped"]).agg({"RIDAGEYR": [np.mean, np.std, np.size]})#.unstack()

RIDAGEYR               
                          mean       std size
SMQ020x Age_grouped                          
No      (18, 28]     23.443491  2.996779  699
        (28, 38]     33.316695  2.835650  581
        (38, 48]     43.595960  2.818582  594
        (48, 58]     53.125786  2.796476  477
        (58, 68]     63.282851  2.750437  449
        (68, 78]     73.194245  2.843392  278
Yes     (18, 28]     23.863281  2.754116  256
        (28, 38]     33.576720  2.883353  378
        (38, 48]     43.650943  2.889257  318
        (48, 58]     53.739899  2.703120  396
        (58, 68]     63.154639  2.925876  485
        (68, 78]     72.784314  2.844474  306

In [320]:
# insert your code here

da.groupby(["SMQ020x", "Age_grouped"]).agg({"RIDAGEYR": [np.mean, np.std, np.size]}).unstack()

RIDAGEYR                                                         \
                  mean                                                          
Age_grouped   (18, 28]   (28, 38]   (38, 48]   (48, 58]   (58, 68]   (68, 78]   
SMQ020x                                                                         
No           23.443491  33.316695  43.595960  53.125786  63.282851  73.194245   
Yes          23.863281  33.576720  43.650943  53.739899  63.154639  72.784314   

                                                                         \
                  std                                                     
Age_grouped  (18, 28]  (28, 38]  (38, 48]  (48, 58]  (58, 68]  (68, 78]   
SMQ020x                                                                   
No           2.996779  2.835650  2.818582  2.796476  2.750437  2.843392   
Yes          2.754116  2.883353  2.889257  2.703120  2.925876  2.844474   

                                                                   
                size                                               
Age_grouped (18, 28] (28, 38] (38, 48] (48, 58] (58, 68] (68, 78]  
SMQ020x                                                            
No               699      581      594      477      449      278  
Yes              256      378      318      396      485      306

__Q6b.__ Does it appear that uncertainty about the mean age of smokers, or uncertainty about the mean age of non-smokers contributed more to the uncertainty for the mean difference that we are focusing on here?